### PR TITLE
333 implement contexts

### PR DESCRIPTION
**Pull request by callumskeet**
Solved issue #336. I've also added functionality to import environment variables from configuration files.

### CODE REVIEW

File: PyOpenWorm/configure.py [hunk 1]
```diff
@@ -134,6 +134,10 @@ def open(cls, file_name):
                         Requirement.parse('PyOpenWorm'),
                         value)
                     d[k] = value
+                if value.startswith("$"):
+                    from os import environ
+                    value = environ['PUBMED_API_KEY']
```
Comment (mwatts15) [inline PyOpenWorm/configure.py]: If we're adding a mechanism for resolving env-vars in the conf, can we make it generic and add some validation (i.e., only alphanumerics and '_', starting with a letter or '_')}

Comment (mwatts15) [reply PyOpenWorm/configure.py]: Also, please account for the var not being defined. I typically assume environment variables may not be defined, so defaulting to `None` is acceptable.

**Comment by callumskeet** [comment]
@mwatts15 I've updated with your recommendations

File: PyOpenWorm/configure.py [hunk 2]
```diff
@@ -134,6 +134,24 @@ def open(cls, file_name):
                         Requirement.parse('PyOpenWorm'),
                         value)
                     d[k] = value
+                if value.startswith("$"):
+                    import re
+                    from os import environ
```
Comment (mwatts15) [inline PyOpenWorm/configure.py]: Generally, I prefer to put imports up top unless there's a dependency issue.

File: PyOpenWorm/configure.py [hunk 3]
```diff
@@ -134,6 +134,24 @@ def open(cls, file_name):
...
+                            value = environ[value]
+                        else:
+                            msg = ("'%s' is an invalid env-var name\n"
+                                  "Env-var names must be alphnumeric "
+                                  "and start with either a letter or '_'" % value)
+                            raise ValueError(msg)
+                    except ValueError:
+                        raise
+                    except Exception:
+                        value = None
```
Comment (mwatts15) [inline PyOpenWorm/configure.py]: I never like to see Exception swallowed like this. I think, here, the most likely other exception is going to be a programmer error, so it should propagate, but should, at least, be logged.

**Comment by mwatts15** [comment]
Thanks, @callumskeet! Good work. I merged as-is, but please see my comments.

**Pull request merged**